pkg/obfuscator/domain.go

```diff
@@ -0,0 +1,109 @@
+package obfuscator
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+)
+
+type domainObfuscator struct {
+	ReplacementReporter
+	hostPattern          *regexp.Regexp
+	domainCount          int
+	lock                 sync.Mutex
+	tlds                 map[string]struct{}
+	mainDomains          map[string]string
+	ipPattern            *regexp.Regexp
+	fileExtensionPattern *regexp.Regexp
+}
+
+func (d *domainObfuscator) FileName(s string) string {
+	// if the filename has an extension omit it from obfuscation
+	// this is still error-prone because the last part of the domain name could still be mistaken for an extension
+	matches := d.fileExtensionPattern.FindStringSubmatch(s)
+	if len(matches) == 3 && matches[2] != "" {
+		return fmt.Sprintf("%s.%s", d.findDomains(matches[1]), matches[2])
+	}
+	return d.findDomains(s)
+}
+
+func (d *domainObfuscator) findDomains(input string) string {
+	domains := d.hostPattern.FindAllString(input, -1)
+
+	output := input
+	for _, domain := range domains {
+		// if this is an IP address then do nothing
+		if d.ipPattern.MatchString(domain) {
+			continue
+		}
+		output = strings.ReplaceAll(output, domain, d.obfuscatedDomain(domain))
+	}
+	return output
+}
+
+func (d *domainObfuscator) Contents(s string) string {
+	return d.findDomains(s)
+}
+
+func (d *domainObfuscator) obfuscatedDomain(domain string) string {
+	d.lock.Lock()
+	defer d.lock.Unlock()
```

Comment:
any reason for the locking yet? let's leave it out for now, the domainCount increment we can also do atomically without locking

Reply:
Removed.